lib/Zonemaster/Backend/DB/SQLite.pm

```diff
@@ -164,97 +165,134 @@ sub create_new_batch_job {
 
     die "You can't create a new batch job, job:[$batch_id] started on:[$creaton_time] still running \n" if ( $batch_id );
 
-    my ( $new_batch_id ) = $self->dbh->selectrow_array(
-        "INSERT INTO batch_jobs (username) VALUES(" . $self->dbh->quote( $username ) . ") RETURNING id" );
+    $self->dbh->do("INSERT INTO batch_jobs (username) VALUES(" . $self->dbh->quote( $username ) . ")" );
+    my ( $new_batch_id ) = $self->dbh->sqlite_last_insert_rowid;
 
     return $new_batch_id;
 }
 
 sub create_new_test {
-    my ( $self, $domain, $test_params, $minutes_between_tests_with_same_params, $batch_id ) = @_;
-    my $result;
+    my ( $self, $domain, $test_params, $minutes, $batch_id ) = @_;
 
-    my $priority = 10;
-    $priority = $test_params->{priority} if (defined $test_params->{priority});
-    
-    my $queue = 0;
-    $queue = $test_params->{queue} if (defined $test_params->{queue});
+    my $dbh = $self->dbh;
 
     $test_params->{domain} = $domain;
-    my $js = JSON::PP->new;
-    $js->canonical( 1 );
+    my $js                             = JSON::PP->new->canonical;
     my $encoded_params                 = $js->encode( $test_params );
     my $test_params_deterministic_hash = md5_hex( $encoded_params );
+    my $result_id;
+
+    my $priority = $test_params->{priority};
+    my $queue = $test_params->{queue};
+
+    # Search for recent test result with the test parameters
+    my ( $recent_id, $recent_hash_id ) = $dbh->selectrow_array(
+        "SELECT id, hash_id FROM test_results WHERE params_deterministic_hash = ? AND test_start_time > DATETIME('now', '-$minutes minutes')",
+        undef,
+	$test_params_deterministic_hash,
+    );
+
+    if ( $recent_id ) {
+        # A recent entry exists, so return its id
+        if ( $recent_id > $self->config->force_hash_id_use_in_API_starting_from_id() ) {
+            $result_id = $recent_hash_id;
+        }
+        else {
+            $result_id = $recent_id;
+        }
+    }
+    else {
+	my $hash_id; # To hold unique id
+	for (1..5) { # 5 trials or else we have a full database or a bug.
+	    $hash_id = substr(md5_hex(time().rand()), 0, 16);
+	    my ( $i ) = $self->dbh->selectrow_array( "SELECT hash_id FROM test_results WHERE hash_id=?", undef, $hash_id );
+	    last unless $i; # Expected to be empty
+	    $hash_id = ''; # Discard since already in use
+	}
+	die 'Cannot create unique hash_id in create_new_test()' unless $hash_id;
```

Comment:
On the contrary, the commit description says that it resolves a bug besides some readability improvements.

No, if two processes will check for the same non-existent hash_id, both will be able to add it since there is no locking of the database. To work-around that would require greater changes. The previous code did not check if the the hash_id was already used.

As far as I can see, the hash_id is not guaranteed to be unique by MySQL (MariaDB) or PostgreSQL engine, so be on the same level, the check should be removed. Just create the hash_id and hope that it is unique.

Do you mean why hash_id is created by the Perl code and not by the database engine?

Reply:
I was completely convinced that `hash_id` was unique! Why isn't it unique? Isn't it supposed to be an identifier?

I didn't think of looking for triggers in the database definition. I thought this was some new invention. Triggers are pretty sneaky. I'm not sure how to best document them. I guess the SQLite module (which doesn't use them) is the wrong place though.

Reply:
`hash_id` is the acting unique identifier, but maybe it was too complicated to make it strictly unique.

Reply:
Updated with a commit that removes the extra check of hash_id to come in line with the other database engines, and also explains why hash_id is created here (but not for the other database engines).